Rewrite in AngularJS 1.X using TypeScript.

CtrlC-Root opened the pull request:
This pull request is a complete rewrite of all existing functionality using AngularJS 1.X and TypeScript. This is a work in progress and there are more things to implement before we can make a release:

* Automatically refresh the view after uploading a file using an HTML form.
* Integrate DropZone as an alternative to using HTML forms directly.
* Ensure all errors are caught and displayed to the user.
* Write unit and integration tests.
* Set up Travis CI to run tests and create builds.
* Things I missed?

Here is what I'd like feedback on specifically:

* Structure and organization of the project. Are there better (more idomatic perhaps) ways to structure AngularJS projects? Should I use namespaces, modules, or both?
* HTML and CSS structure. Is the HTML I used clear and straight-forward? Is it flexible enough to allow appropriate restyling?
* Core functionality. This includes browsing the bucket, creating folders, uploading files, deleting folders, deleting files, downloading files, and downloading specific file versions.
* User interface. Is it intuitive? Is the layout good?

I will end up expanding the `README.md` based on the feedback I get. For now, to test this, you will need an S3 bucket with an appropriate CORS configuration (see a very permissive example below).

```
<?xml version="1.0" encoding="UTF-8"?>
<CORSConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
<CORSRule>
    <AllowedOrigin>*</AllowedOrigin>
    <AllowedMethod>GET</AllowedMethod>
    <AllowedMethod>PUT</AllowedMethod>
    <AllowedMethod>POST</AllowedMethod>
    <AllowedMethod>DELETE</AllowedMethod>
    <MaxAgeSeconds>3000</MaxAgeSeconds>
    <AllowedHeader>*</AllowedHeader>
</CORSRule>
</CORSConfiguration>
```

CtrlC-Root committed:
Remove existing files.

This provides a clean slate to start a new implementation.

CtrlC-Root committed:
Use yeoman fountain generator to create new AngularJS project.

This provides a working project structure out of the box. I chose to use
AngularJS 1, Webpack, TypeScript, and CSS.

CtrlC-Root committed:
Remove content generated by yeoman fountain generator.

We will not be using any of the HTML, CSS, or JS content provided by the
yeoman generator as examples.

CtrlC-Root committed:
Rename the angular module from app to s3commander.

This makes the project structure easier to understand especially given
the default app component that was present.

CtrlC-Root committed:
Replace a TypeScript configuration parameter.

This fixes a depcrecation warning visible when running any gulp tasks
that use TypeScript.

CtrlC-Root committed:
Update the README.md with usage instructions.

This makes it easy for newcomers to get started running and building the
application.

CtrlC-Root committed:
Remove app component and default router config.

We want to create our own components and we most likely won't need
client-side routing since s3c is designed to be embedded.

CtrlC-Root committed:
Add initial Bucket component.

The Bucket component will be the top-level component used to display
bucket contents.

CtrlC-Root committed:
Add initial S3 backend and refactor structure.

Create an initial backend interface and S3 implementation. Move the
Bucket model into the common folder with the backends because that's
where it'll be used most. Update the bucket component imports.

CtrlC-Root committed:
Add a reference link to a best practices guide to the README.md.

This guide is useful for using AngularJS 1.X with TypeScript.

CtrlC-Root committed:
Fix a syntax error in the AmazonS3Backend class.

Classes implemnent interfaces, they do not extend them.

CtrlC-Root committed:
Modify bucket controller to use appropriate backend.

The BucketController will create an instance of an appropriate IBackend
implementation and use it to retrieve the requested Bucket.

CtrlC-Root committed:
Add a link to the AngularJS style guide to the README.md.

The official style guide is useful for writing correctly formatted and
idiomatic code.

CtrlC-Root committed:
Compute and display the Bucket base URL.

This demonstrates how component bindings can influence the backend
through the controller. It will also be used down the line for S3 API
calls.

CtrlC-Root committed:
Add the AWS SDK to the project dependencies.

Now we can use the SDK instead of writing our own code to make HTTP API
calls (and sign them).

CtrlC-Root committed:
Modify the Amazon backend to use the AWS SDK.

The Amazon S3 backend now uses the AWS SDK to retrieve bucket details.
The Bucket component has also been updated to work with the modified
backend which returns Promise objects.

CtrlC-Root committed:
Remove unused IBucketScope interface.

We no longer use a local scope in the BucketController.

CtrlC-Root committed:
Fix a documentation comment for a controller attribute.

CtrlC-Root committed:
Fill in the initial structure of the bucket template.

Put some TODO comments in the template for where future components will
go after they are implemented.

CtrlC-Root committed:
Add initial Path class implementation.

The Path class abstracts away working with bucket paths.

CtrlC-Root committed:
Modify the Path class to omit leading slashes in paths.

This makes it easier to work with S3, B2, and other web-based object
storage services because all of their paths are relative to a
proto://bucket.hostname:port/ bucket URL.

CtrlC-Root committed:
Implement fetching bucket objects.

The name Object is already in use in TypeScript so I named the class
StorageObject.

CtrlC-Root committed:
Retrieve folder and file objects in the bucket controller.

Also rename Bucket to StorageBucket so it's consistent with
StorageObject. Oh whatever, namespaces are what?

CtrlC-Root committed:
Implement and use File and Folder components.

Modify the bucket component template to display folders and files using
the new components.

CtrlC-Root committed:
Show the path breadcrumbs in the bucket template.

This uses the current working path to display breadcrumbs in the bucket
details section of the bucket template.

CtrlC-Root committed:
Implement clone, equals, push, and pop methods for the Path class.

This allows easier manipulation of Path objects.

CtrlC-Root committed:
Fix AmazonS3Backend storage object paths.

The S3 backend now creates correct paths for storage objects. It also
ignores the folder object for the given prefix so it doesn't show up in
the files array.

CtrlC-Root committed:
Add navigate and delete events to the folder controller and template.

Prepare for exposing the methods as component bindings to the parent.

CtrlC-Root committed:
Implement basic navigation in bucket controller.

Users can now navigate to folders or to the parent folder.

CtrlC-Root committed:
Implement deleting folders.

The code deletes all folder and file objects with a path that has the
top-level folder path as its prefix.

CtrlC-Root committed:
Implement deleting files.

We can use the same backend function to delete a file as we do to delete
folders.

CtrlC-Root committed:
Improve the overall layout and style.

Style the sections and buttons to be more equally spaced out and look
better. This makes working on this project easier on my eyes.

CtrlC-Root committed:
Implement creating folders.

Use empty objects with keys that end in a trailing slash to represent
folders.

CtrlC-Root committed:
Rename StorageBucket class to Bucket.

This is in preparation for rewriting the backend API to use File and
Folder objects instead of StorageObject objects.

CtrlC-Root committed:
Replace StorageObject with File and Folder classes.

Reimplement the IBackend interface and AmazonS3Backend class to use File
and Folder objects based on an IBUcketObject interface instead of the
StorageObject class. This provides a much cleaner and more type-safe API
than before.

CtrlC-Root committed:
Implement a CSS only loading spinner.

This will probably be overriden by the end user anyways but it's nicer
to look at than the "LOADING" text I had there before.

CtrlC-Root committed:
Fix the Path.clone() method to create a proper clone.

The clone method was relying on Path.toString() which is not a perfect
representation of the Path object's internal state. Explicitly copying
the member attributes is better.

CtrlC-Root committed:
Implement downloading files.

Right now we don't explicitly make the file <a> tags links themselves
but we simulate clicking on one. Once support for file versions is added
it may make sense to change this implementation.

CtrlC-Root committed:
Do not pass in unused $timeout service to the bucket controller.

We're not using the $timeout service at all in the bucket controller so
get rid of it.

CtrlC-Root committed:
Add a file upload form and improve overall styling.

The file upload form is not wired to anything yet. I tried to style the
label to look just like the buttons we already use.

CtrlC-Root committed:
Improve the visibility of individual folder and file rows.

Set a background color and adjust the spacing between rows to make them
more easily distinguishable and visisble to the end user.

CtrlC-Root committed:
Implement backend support for retrieving file versions.

Add a new method getFileVersions() to the IBackend interface and
AmazonS3Backend implementation for retrieving file versions and delete
markers. This is paving the way for downloading specific file versions.

CtrlC-Root committed:
Add button to load and display file versions.

This allows the user to request file versions for specific files when
they want to see them as opposed to loading them all up front.

CtrlC-Root committed:
Document the IFileVersion interface.

Document the interface and its' members.

CtrlC-Root committed:
Refactor IFolderContents interface into its own file.

Now all interfaces and classes are in separate files which makes the
source code more consistently organized.

CtrlC-Root committed:
Pre-emptively generate file download links instead of on-demand.

Remove the backend support for explicitly generating download links.
When retrieving files in a folder or file versions we automatically
generate the download links where appropriate and include them in the
response.

CtrlC-Root committed:
Fix a TypeScript warning.

Holy moly, sometimes TypeScript really wants what it wants and I'm tired
of seeing the warning messages.

CtrlC-Root committed:
Correct a TypeScript function type definition.

Just because.

CtrlC-Root committed:
Only show file versions button when parent bucket is versioned.

We only allow users to request file versions if the parent bucket is
versioned. When the bucket is not versioned the button is hidden.

CtrlC-Root committed:
Configure the s3commander module to whitelist form URLs.

By default AngularJS doesn't allow you to interpolate form URLs unless
they are whitelisted in the module's security policy. I added two URLs
for the AWS S3 API to the whitelist.

CtrlC-Root committed:
Add backend support for getting file upload settings.

We can create pre-signed POST requests using the S3 API and we can use
those settings to upload files using HTML forms.

CtrlC-Root committed:
Implement a file upload form using directives and components.

Keeping complicated logic split into smaller components makes things
easier to understand in the long run. Also all DOM manipulation logic is
kept inside of directives and not in controllers.

CtrlC-Root committed:
Update a comment to actually describe the motivation behind it.

CtrlC-Root committed:
Submit the file upload form when the user selects a valid file.

This is implemented by exposing the form.submit() method through a
directive.

CtrlC-Root committed:
Add a link to an AngularJS cheat sheet to the README.md.

CtrlC-Root committed:
Minor code cleanup.

Use the .on() method instead of the deprecated .bind() method. Use a
proper TypeScript type instead of any for casting an HTML form element.

CtrlC-Root committed:
Create an S3 session component.

Now we can easily test the bucket functionality without modifying the
source code.

CtrlC-Root committed:
Format file version timestamps using the current locale.

Now they show up as user readable date / time string.

CtrlC-Root commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Clicking "Delete" on directories in S3Commander deletes the objects in S3, but fails to display the directory as deleted to the user](https://trello.com/c/GFPvy9uw/2070-clicking-delete-on-directories-in-s3commander-deletes-the-objects-in-s3-but-fails-to-display-the-directory-as-deleted-to-the-use)

ericamador commented on README.md:
```diff
@@ -2,19 +2,41 @@
-TODO
+* Tools
+  * [Yeoman](http://yeoman.io/)
+  * [Yeaoman FountainJS Generator](http://fountainjs.io/)
```
Typo in "Yeoman"?

CtrlC-Root replied:
Yup, good catch haha.

ericamador commented on src/common/AmazonS3Backend.ts:
```diff
@@ -2,15 +2,26 @@ import {Bucket} from './Bucket';
-    return new Bucket(name);
+    // when the bucket name contains a period ('.') we can't use https
+    // because it treats part of the bucket name as a subdomain and breaks SSL
+    let protocol = name.includes('.') ? 'http' : 'https';
```
Should there be a note in the README warning users of the fact that SSL can't be used when a period appears in a bucket name?

CtrlC-Root replied:
I actually think the AWS SDK uses a different endpoint for API calls so it doesn't have the same limitation. The exception to this is when I override it for the HTML file uploads. I'll add a note to the README.md for that.

ericamador commented on src/bucket/BucketController.ts:
```diff
@@ -42,7 +42,7 @@ export class BucketController {
 
   /**
-   * TODO.
+   * Error encountered running background operation.
```
May as well just squash this into 957f9d.

ericamador commented:
Looks great. Just a couple other minor notes:

- I had to lookup how to run tslint (`tslint -c tslint.json 'src/**/*.ts'`). Maybe that should be included in the README?
- The pep8/flake8 reference in CONTRIBUTING.md should probably be updated to reference tslint instead.

ericamador commented:
tslint reports 4 remaining errors:

```
ERROR: src/common/IUploadConfig.ts[14, 17]: missing whitespace
ERROR: src/s3session/S3SessionController.ts[31, 7]: comment must start with lowercase letter
ERROR: src/s3session/S3SessionController.ts[38, 7]: comment must start with lowercase letter
ERROR: src/s3session/S3SessionController.ts[39, 7]: comment must start with lowercase letter
```

CtrlC-Root commented:
@amadornimbis Regarding `tslint` I think a better approach would be to add a `gulp` task to run it for you. Maybe `gulp lint`? I'll fix the first error but the remaining 3 are caused by `// TODO` comments which will be removed in a future PR. Actually, it kind of bothers me those are issues because of uppercase versus lowercase. I would much rather have a warning regarding `TODO` and `XXX` than an error because my comment isn't lowercase, so maybe I'll look into changing it's configuration to do that.

Regarding the CONTRIBUTING.md file, I plan to update it once I add tests in a future PR so I can reference the entire testing procedure.

CtrlC-Root committed:
Minor formatting and README.md cleanup.

Fix a whitespace lint warning in IUploadConfig. Fix a typo in the
README.md. Add a Quirks section to the README.md and add an item that
mentions HTML file uploads only work when your bucket name is HTTPS
compatible.

CtrlC-Root committed:
Configure tslint to ignore TODO and XXX in single line comments.

By default tslint complains if the first character in a asingle line
comment is not lowercase. This means TODO and XXX comments generate
warnings (for uppercase) which is annoying. Now tslint will ignore those
words when checking the comment case.

CtrlC-Root committed:
Add a gulp task to run the TypeScript linter.

Now we can run this separately from gulp serve for testing purposes.

CtrlC-Root committed:
Fix the gulp tslint task to correctly report errors.

Apparently there's a breaking change in tslint 5.x from 4.x that causes
gulp-tslint not to work correctly. After upgrading the package to 5.9.1
everything else seems to still work correctly and it fixes this issue.

CtrlC-Root committed:
Add a missing whitespace to satisfy tslint.

CtrlC-Root committed:
Validate folder names before creating folders.

Folder names must match a regular expression and cannot be empty.
Initialize the default folder name to an empty string in the constructor
to avoid unfortuante JavaScript typecast to 'undefined'. Improve overall
error reporting and handling in the bucket controller.

CtrlC-Root commented:
Ready for another look.

* Updated README.md to mention HTTPS and HTML file upload quirk.
* Resolved the tslint warnings by fixing the code and adjusting the configuration of the linter.
* Added a `gulp tslint` task and updated the README.md file to mention it.

CtrlC-Root committed:
Add a Vagrant VM for development.

This makes it easy to develop on machines that do not have access to the
latest versions of Node.js or NPM.

CtrlC-Root committed:
Fix the S3 backend so file uploads to GovCloud regions work.

Move the code that overwrites the file upload POST url into the
AmazonS3Backend class where it belongs because it's S3 specific. Modify
the code to reuse the provided hostname instead of hard-coding
s3.amazonaws.com which does not work for GovCloud regions. Add a new
rule to the URL interpolation whitelist to allow sub-region S3 URLs.

mliudev commented on .yo-rc.json:
```diff
@@ -0,0 +1,17 @@
+      "namespace": "fountain-angular1:app",
+      "_": [],
+      "sample": "techs",
+      "router": "uirouter"
```
Not using routing so this should probably go.

CtrlC-Root replied:
Good catch, I'll remove it.

mliudev commented on package.json:
```diff
@@ -0,0 +1,92 @@
+  "license": "SEE LICENSE IN LICENSE.txt",
+  "dependencies": {
+    "angular": "^1.6.2",
+    "angular-ui-router": "1.0.0-beta.3",
```
Doesn't look like we're using this.

CtrlC-Root replied:
Yup, needs to be removed.

mliudev commented on package.json:
```diff
@@ -0,0 +1,92 @@
+  "devDependencies": {
+    "@types/angular": "^1.6.6",
+    "@types/angular-mocks": "^1.5.9",
+    "@types/angular-ui-router": "^1.1.36",
```
Not using this I think.

CtrlC-Root replied:
Yup.

mliudev commented on tsconfig.json:
```diff
@@ -0,0 +1,25 @@
+    "noImplicitAny": false,
+    "typeRoots": ["node_modules/@types/"],
+    "types": [
+      "angular-ui-router",
```
Not using this either.

CtrlC-Root replied:
Good lord.

mliudev commented on src/common/AmazonS3Backend.ts:
```diff
@@ -1,7 +1,7 @@
 import {IBackend} from './IBackend';
 
-export class AmazonS3Backend extends IBackend {
+export class AmazonS3Backend implements IBackend {
```
We generally rebase these commits into the one where the error was originally committed. This reduces the number of commits and helps the reviewer comprehend the code more easily.

mliudev commented on src/bucket/BucketController.ts:
```diff
@@ -29,6 +47,15 @@ export class BucketController {
+        throw ('Unknown backend: ' + this.backendName);
+    }
+
+    this.bucket = this.$scope.bucket = this.backend.getBucket(this.bucketName);
```
Why is it that we set a private bucket as well as a scope bucket? This seems like a duplication of the object.

mliudev commented on src/bucket/BucketController.ts:
```diff
@@ -24,38 +24,76 @@ export class BucketController {
+      throw `Unknown backend: ${this.backendName}`;
     }
 
-    this.bucket = this.$scope.bucket = this.backend.getBucket(this.bucketName);
```
This negates my question about keeping two buckets.

mliudev commented on src/common/Path.ts:
```diff
@@ -0,0 +1,56 @@
+  /**
+   * Check if this path refers to a folder.
+   */
+  public isFolder(): boolean {
```
What do you think about making this into a property with a getter and setter?
Check out the accessors section here: https://www.typescriptlang.org/docs/handbook/classes.html.

CtrlC-Root replied:
Hmm, I think it's overkill for this property. I only want the internal implementation of the class to ever set it and end-users of the class to get it.

mliudev replied:
👍

mliudev commented on src/common/Path.ts:
```diff
@@ -46,6 +46,12 @@ export class Path {
    * Get the string representation.
    */
   public toString(): string {
+    // corner case: empty path
```
I'm ambivalent about squashing this into the previous commit, but it'd be a little cleaner if you did.

mliudev commented on src/bucket/BucketController.ts:
```diff
@@ -0,0 +1,254 @@
+        this.awsAccessKeyId,
+        this.awsSecretAccessKey);
+    } else {
+      throw `Unknown backend: ${this.backendName}`;
```
This gitbook recommends against throwing string errors: https://basarat.gitbooks.io/typescript/content/docs/types/exceptions.html. See the "Always use Error" section.

CtrlC-Root replied:
I must have missed that, thanks for pointing it out. I will update all of my errors to use the `new Error(...)` approach instead.

mliudev commented on src/bucket/BucketController.ts:
```diff
@@ -75,18 +94,44 @@ export class BucketController {
+      this.bucket = bucket;
+      return this.backend.getObjects(bucket, this.workingPath);
+    }).then((objects: StorageObject[]) => {
+      function compareObjectNames (a: StorageObject, b: StorageObject) {
```
Is this function created every time this promise is called? Why not have this just be a regular function in this controller scope?

CtrlC-Root replied:
I think I did move it up to the controller scope in a later commit.

mliudev commented on src/file/file.html:
```diff
@@ -1,5 +1,5 @@
   <!-- TODO: icon -->
-  <a>{{ $ctrl.file.path.name() }}</a>
-  <button class="s3c-item-button">Delete</button>
+  <a ng-click="$ctrl.onDownload()">{{ $ctrl.file.path.name() }}</a>
```
Does this line belong in this commit?

mliudev commented on src/bucket/BucketController.ts:
```diff
@@ -12,8 +12,7 @@ export class BucketController {
   public static $inject = [
-    '$rootScope',
-    '$timeout'
+    '$rootScope'
```
This commit can be squashed into the previous one.

mliudev commented on src/index.css:
```diff
@@ -9,28 +9,24 @@ html, body {
  */
 
 div.s3c-bucket {
-  border: 1px dashed #ccc;
```
Typo in the commit 'visisble' no action needed unless you really want to.

mliudev commented on src/common/AmazonS3Backend.ts:
```diff
@@ -149,11 +143,19 @@ export class AmazonS3Backend implements IBackend {
+            Bucket: bucket.name,
+            Key: file.getPath().toString(),
+            VersionId: versionData.VersionId,
+            Expires: 900
```
This seems like a rather arbitrary expiration time. If a user left this window open for over 900 ticks would they no longer be able to download this version?

CtrlC-Root replied:
Yup. We should probably implement an auto-refresh in a future version that indirectly leads to generating new links. Or maybe just remove it? It's really not a big deal when you're using temporary credentials which I believe we're planning on moving to. If you're straight-up using your IAM access and secret keys though I think it's dangerous to not set an expiration time. A rogue plugin could grab that link and hold on it or something.

mliudev replied:
OK, either of these solutions sound good to me.

mliudev commented on src/file/FileController.ts:
```diff
@@ -69,9 +69,12 @@ export class FileController {
-    return new Promise(function (resolve, fail) {
-      fail('not implemented');
+    // overriden by binding but TypeScript requires we return something
+    return new Promise(function (
```
Why use function here instead of fat arrow notation like you have before? I'm more interested if there are any differences other than preference.

CtrlC-Root replied:
Arrow notation does not change the value of `this` which is almost always what you want. I'm just not used to the syntax though so I sometimes forget to use it. I believe I changed all the old-style functions into arrow functions in a later commit. If not, point them out and I'll change them.

mliudev commented on src/file/FileController.ts:
```diff
@@ -71,7 +71,7 @@ export class FileController {
     // overriden by binding but TypeScript requires we return something
     return new Promise(function (
-      resolve: (data: any) => void,
+      resolve: (data: IFileVersion[]) => void,
```
Squash.

mliudev commented on src/bucket/BucketComponent.ts:
```diff
@@ -2,11 +2,11 @@ import {BucketController} from './BucketController';
-    awsRegion: '@awsRegion',
-    awsAccessKeyId: '@awsAccessKeyId',
-    awsSecretAccessKey: '@awsSecretAccessKey'
+    backendName: '=backend',
```
This looks like it breaks the component guide convention to never use '=' bindings: https://docs.angularjs.org/guide/component.

```
Inputs should be using < and @ bindings. The < symbol denotes one-way bindings which are available since 1.5.
The difference to = is that the bound properties in the component scope are not watched, which means if you 
assign a new value to the property in the component scope, it will not update the parent scope. Note however, 
that both parent and component scope reference the same object, so if you are changing object properties or 
array elements in the component, the parent will still reflect that change. The general rule should therefore be to 
never change an object or array property in the component scope. @ bindings can be used when the input is a 
string, especially when the value of the binding doesn't change.

bindings: {
  hero: '<',
  comment: '@'
}
```

CtrlC-Root replied:
Hmm, there was a reason I used `=` instead of `@` here and I don't remember it. I'll have to change it to `@` and see if everything still works.

mliudev commented on README.md:
```diff
@@ -30,11 +30,15 @@ Create a release build.
 * Tools
   * [Yeoman](http://yeoman.io/)
-  * [Yeaoman FountainJS Generator](http://fountainjs.io/)
+  * [Yeaman FountainJS Generator](http://fountainjs.io/)
```
Yeoman?

mliudev commented on README.md:
```diff
@@ -46,7 +46,7 @@ gulp tslint
 * Tools
   * [Yeoman](http://yeoman.io/)
-  * [Yeaman FountainJS Generator](http://fountainjs.io/)
+  * [Yeoman FountainJS Generator](http://fountainjs.io/)
```
You caught it.

mliudev reviewed (approved):
The functionality of the code looks good. I make a variety of comments, most of which are minor, as well as several questions. No major changes are requested.

mliudev commented on src/bucket/BucketController.ts:
```diff
@@ -0,0 +1,254 @@
+        return this.backend.getContents(bucket, this.currentFolder);
+      })
+      .then((contents: IFolderContents) => {
+        function compareObjectNames (a: IBucketObject, b: IBucketObject) {
```
This is what I was referring to in the outdated commit about refactoring this function out instead of instantiating it here every time this promise runs. Doubt that it's a hit on performance. I only bring this up because you use a similar pattern in the File module but it's declared outside the promise.

mliudev commented on src/file/FileController.ts:
```diff
@@ -0,0 +1,108 @@
+   */
+  public onLoadVersions(): Promise<IFileVersion[]> {
+    // overriden by binding but TypeScript requires we return something
+    return new Promise(function (
```
Here is where you don't use fat arrow notation, but since you don't reference 'this' in the function there should be no difference.

CtrlC-Root committed:
Fix a bug caused by cancelling the file selection dialog.

The FileInputDirective was not correctly handling the case where the
user opened the file dialog and then closed it by clicking the Cancel
button without selecting a file. Consequently the UploadFormController
was not handling that case either. This lead to a pretty harmless JS
error in the browser console when encountered.

CtrlC-Root committed:
Remove remaining references to Angular UI router package.

We don't use the routing functionality in Angular at all.

CtrlC-Root committed:
Throw Error objects instead of strings.

This follows best practices for TypeScript.

CtrlC-Root committed:
Always use arrow function notation.

This keeps the code more consistent.
